src/Analysis/Elaboration.hs

```diff
@@ -24,69 +25,81 @@ runElab :: Elab a -> (a, Subst PartialType)
 runElab = fst . flip runFresh (Name 0) . flip runReaderT mempty . flip runStateT mempty
 
 elaborate :: Term -> Elab PartialElabTerm
-elaborate (Fix (Abs n b)) = do
+elaborate term = do
+  term' <- infer term
+  subst <- get
+  let ty :< tm = substitute subst term'
+  pure (generalize ty :< tm)
+
+infer :: Term -> Elab PartialElabTerm
+infer (Fix (Abs n b)) = do
   t <- fresh
-  b' <- local (envExtend n t) (elaborate b)
+  b' <- local (envExtend n t) (infer b)
   pure ((tvar t .-> extract b') :< Abs n b')
-elaborate (Fix (Var name)) = do
+infer (Fix (Var name)) = do
   env <- ask
   pure (maybe (Pure (FreeVariable name)) tvar (envLookup name env) :< Var name)
-elaborate (Fix (App f a)) = do
+infer (Fix (App f a)) = do
   t <- fresh
-  a' <- elaborate a
+  a' <- infer a
   f' <- check f (extract a' .-> tvar t)
   pure (tvar t :< App f' a')
-elaborate (Fix (Rec n b)) = do
+infer (Fix (Rec n b)) = do
   t <- fresh
-  local (envExtend n t) (elaborate b)
-elaborate (Fix Expr.Unit) = pure (unitT :< Expr.Unit)
-elaborate (Fix (Pair fst snd)) = do
-  fst' <- elaborate fst
-  snd' <- elaborate snd
+  local (envExtend n t) (check b (tvar t))
```

Comment:
This fixed a bug where we weren’t unifying the type against the result, and thus we got weirdly incorrect types.